File: src/test/scala/com/feedzai/cosytest/junit5/IntegrationSpec.java
package com.feedzai.cosytest.junit5;

import com.feedzai.cosytest.wrapper.DockerComposeJavaSetup;
import com.feedzai.cosytest.wrapper.SetupManager;
import org.junit.Assert;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.RegisterExtension;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.file.Paths;
import java.util.Collections;
import java.util.HashMap;

import static org.hamcrest.Matchers.contains;
import static org.hamcrest.Matchers.containsInAnyOrder;


public class IntegrationSpec {

    private static final Logger logger = LoggerFactory.getLogger(IntegrationSpec.class);

    private static final DockerComposeJavaSetup dockerSetup;

    static {
        dockerSetup = new DockerComposeJavaSetup(
            "junit5test",
            Collections.singletonList(Paths.get("src", "test", "resources", "docker-compose-junit5.yml")),
            Paths.get("").toAbsolutePath(),
            new HashMap<>()
        );
    }

    @RegisterExtension
    static DockerComposeExtension extension = new DockerComposeExtension(SetupManager.builder(dockerSetup).build());

    @BeforeAll
    public static void init() {
        logger.info("Running JUnit 5 Integration Spec!");
    }

    @Test
    public void fetchServices() {
        Assert.assertThat(
            dockerSetup.getServices(),
            containsInAnyOrder("container1", "container2", "container3")
        );
    }

    @Test
    public void validateMappedPorts() {
        Assert.assertThat(dockerSetup.getServiceMappedPort("container1", 80), contains("8090"));
        Assert.assertThat(dockerSetup.getServiceMappedPort("container2", 80), contains("8091"));
        Assert.assertThat(dockerSetup.getServiceMappedPort("container3", 80), contains("8092"));
    }

}

File: src/test/scala/com/feedzai/cosytest/scalatest/IntegrationSpec.scala
package com.feedzai.cosytest.scalatest

import java.nio.file.Paths

import com.feedzai.cosytest.core.DockerComposeSetup
import org.scalatest.{FlatSpec, MustMatchers}

class IntegrationSpec extends FlatSpec with DockerComposeTestSuite with MustMatchers {

  override def dockerSetup = Some(
    DockerComposeSetup(
      "scalatest",
      Seq(Paths.get("src", "test", "resources", "docker-compose-scalatest.yml")),
      Paths.get("").toAbsolutePath,
      Map.empty
    )
  )

  behavior of "Scala Test"

  it must "Retrieve all services" in {
    val expectedServices = Set("container1", "container2", "container3")
    dockerSetup.foreach { setup =>
      setup.getServices().toSet mustEqual expectedServices
    }
  }

  it must "Retrieve correct mapped ports for all services" in {
    dockerSetup.foreach { setup =>
      val service1  = setup.getServiceMappedPort("container1", 80)
      service1.size mustBe 1
      service1.head mustEqual "8081"
      val service2  = setup.getServiceMappedPort("container2", 80)
      service2.size mustBe 1
      service2.head mustEqual "8082"
      val service3  = setup.getServiceMappedPort("container3", 80)
      service3.size mustBe 1
      service3.head mustEqual "8083"
    }
  }
}
